src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java

```diff
@@ -144,4 +164,102 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
 
     private static final Logger LOGGER = Logger.getLogger(NaginatorPublisher.class.getName());
 
+    public static class Action extends NaginatorScheduleAction {
+        private final String regexpForRerun;
+        private final boolean rerunIfUnstable;
+        private final boolean checkRegexp;
+        
+        public Action(NaginatorPublisher publisher) {
+            super(publisher.getMaxSchedule(), publisher.getDelay(), publisher.isRerunMatrixPart());
+            this.regexpForRerun = publisher.getRegexpForRerun();
+            this.rerunIfUnstable = publisher.isRerunIfUnstable();
+            this.checkRegexp = publisher.isCheckRegexp();
+        }
+        
+        public String getRegexpForRerun() {
+            return regexpForRerun;
+        }
+        
+        public boolean isRerunIfUnstable() {
+            return rerunIfUnstable;
+        }
+        
+        public boolean isCheckRegexp() {
+            return checkRegexp;
+        }
+        
+        @Override
+        public boolean shouldSchedule(Run<?, ?> run, TaskListener listener, int retryCount) {
+            if ((run.getResult() == Result.SUCCESS) || (run.getResult() == Result.ABORTED)) {
+                return false;
+            }
+            
+            // If we're not set to rerun if unstable, and the build's unstable, return true.
+            if ((!isRerunIfUnstable()) && (run.getResult() == Result.UNSTABLE)) {
+                return false;
+            }
+            
+            // If we're supposed to check for a regular expression in the build output before
+            // scheduling a new build, do so.
+            if (isCheckRegexp()) {
+                LOGGER.log(Level.FINEST, "Got checkRegexp == true");
+                
+                String regexpForRerun = getRegexpForRerun();
+                if ((regexpForRerun != null) && (!regexpForRerun.equals(""))) {
+                    LOGGER.log(Level.FINEST, "regexpForRerun - {0}", regexpForRerun);
+                    
+                    try {
+                        // If parseLog returns false, we didn't find the regular expression,
+                        // so return true.
+                        if (!parseLog(run.getLogFile(), regexpForRerun)) {
+                            LOGGER.log(Level.FINEST, "regexp not in logfile");
+                            return false;
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace(listener
+                                          .error("error while parsing logs for naginator - forcing rebuild."));
+                    }
+                }
+            }
+            
+            return super.shouldSchedule(run, listener, retryCount);
+        }
+
+        @Override
+        public boolean shouldScheduleForMatrixRun(MatrixRun run, TaskListener listener) {
```

Comment:
Done with 55c8d827099553a7d03c40662658c62f36117798.
I'm stranger in `@Nonnull` and `@Nullable` annotations and would you review that?